whipper/program/cdrdao.py

```diff
@@ -2,58 +2,162 @@
+
+    def parse(self, line):
+        cdrdao_m = _BEGIN_CDRDAO_RE.match(line)
+
+        if cdrdao_m:
+            logger.debug("RE: Begin cdrdao toc-read")
+
+        leadout_m = _LEADOUT_RE.match(line)
+
+        if leadout_m:
+            logger.debug("RE: Reached leadout")
+            last_track_m = _LAST_TRACK_RE.match(self.oldline)
+            if last_track_m:
+                self.tracks = last_track_m.group('track')
+
+        track_s = _TRACK_RE.search(line)
+        if track_s:
+            logger.debug("RE: Began reading track: %d" %
+                         int(track_s.group('track')))
+            self.currentTrack = int(track_s.group('track'))
+
+        crc_s = _CRC_RE.search(line)
+        if crc_s:
+            print("Track %d finished, found %d Q sub-channels "
+                  "with CRC errors" % (self.currentTrack,
+                                       int(crc_s.group('channels'))))
+
+        self.oldline = line
+
+
+class ReadTOCTask(task.Task):
     """
-    Return cdrdao-generated table of contents for 'device'.
+    Task that reads the TOC of the disc using cdrdao
     """
-    # cdrdao MUST be passed a non-existing filename as its last argument
-    # to write the TOC to; it does not support writing to stdout or
-    # overwriting an existing file, nor does linux seem to support
-    # locking a non-existant file. Thus, this race-condition introducing
-    # hack is carried from morituri to whipper and will be removed when
-    # cdrdao is fixed.
-    fd, tocfile = tempfile.mkstemp(suffix=u'.cdrdao.read-toc.whipper')
-    os.close(fd)
-    os.unlink(tocfile)
-
-    cmd = [CDRDAO, 'read-toc'] + (['--fast-toc'] if fast_toc else []) + [
-        '--device', device, tocfile]
-    # PIPE is the closest to >/dev/null we can get
-    logger.debug("executing %r", cmd)
-    p = Popen(cmd, stdout=PIPE, stderr=PIPE)
-    _, stderr = p.communicate()
-    if p.returncode != 0:
-        msg = 'cdrdao read-toc failed: return code is non-zero: ' + \
-              str(p.returncode)
-        logger.critical(msg)
-        # Gracefully handle missing disc
-        if "ERROR: Unit not ready, giving up." in stderr:
-            raise EjectError(device, "no disc detected")
-        raise IOError(msg)
-
-    toc = TocFile(tocfile)
-    toc.parse()
-    if toc_path is not None:
-        t_comp = os.path.abspath(toc_path).split(os.sep)
-        t_dirn = os.sep.join(t_comp[:-1])
-        # If the output path doesn't exist, make it recursively
-        if not os.path.isdir(t_dirn):
-            os.makedirs(t_dirn)
-        t_dst = truncate_filename(os.path.join(t_dirn, t_comp[-1] + '.toc'))
-        shutil.copy(tocfile, os.path.join(t_dirn, t_dst))
-    os.unlink(tocfile)
-    return toc
+    description = "Reading TOC"
+    toc = None
+
+    def __init__(self, device, fast_toc=False, toc_path=None):
+        """
+        Read the TOC for 'device'.
+        @param device:  block device to read TOC from
```

Comment:
There should be a blank line before this one. See https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings :)
```suggestion

        @param device:  block device to read TOC from
```